etna/lib/etna/clients/magma/workflows/json_models.rb

```diff
@@ -0,0 +1,359 @@
+          validate_attributes
+        end
+
+        def validate_add_model_data
+          if !is_project?
+            check_key("model #{name}", @raw, 'parent_model_name')
+            check_key("model #{name}", @raw, 'parent_link_type')
+            check_type("model #{name}", @raw, 'parent_link_type', @valid_parent_link_types)
+          end
+          if parent_link_type != Etna::Clients::Magma::ParentLinkType::TABLE
+            check_key("model #{name}", @raw, 'identifier')
+          end
+        end
+
+        def validate_attributes
+          attributes.each do |attribute|
+            @errors += attribute.errors unless attribute.valid?
+          end
+        end
+
+        def to_magma_model(builder)
+          template_builder = builder.build_template
+          template_builder.identifier = identifier
+          template_builder.parent = parent_model_name
+          attribute_builder = template_builder.build_attributes
+          attributes.each do |attribute|
+            attribute.to_magma_model(attribute_builder)
+          end
+
+          # Because the input JSON format doesn't specify the parent
+          #   attribute, we'll need to add those
+          #   in manually.
+          add_parent_attribute(attribute_builder) unless is_project?
+        end
+
+        def add_parent_attribute(builder)
+          builder.build_attribute(parent_model_name).tap do |attribute|
+            attribute.attribute_name = parent_model_name
+            attribute.name = parent_model_name
+            attribute.link_model_name = parent_model_name
+            attribute.attribute_type = Etna::Clients::Magma::AttributeType::PARENT
+            attribute.display_name = prettified_parent_model_name
+            attribute.desc = prettified_parent_model_name
+          end
+        end
+      end
+
+      class JsonAttribute < JsonBase
+        attr_reader :name, :model, :raw
+        def initialize(model, attribute_name, raw)
+          super()
+          @model = model
+          @name = attribute_name.strip
+          @raw = raw
+
+          # NOTE: for input simplicity, I've removed some of the
+          #   link-related types, since we'll try to calculate
+          #   those while parsing the JSON structure.
+          @valid_attribute_types = [
+            Etna::Clients::Magma::AttributeType::STRING,
+            Etna::Clients::Magma::AttributeType::DATE_TIME,
+            Etna::Clients::Magma::AttributeType::BOOLEAN,
+            Etna::Clients::Magma::AttributeType::FILE,
+            Etna::Clients::Magma::AttributeType::FLOAT,
+            Etna::Clients::Magma::AttributeType::IMAGE,
+            Etna::Clients::Magma::AttributeType::INTEGER,
+            Etna::Clients::Magma::AttributeType::LINK,
+            Etna::Clients::Magma::AttributeType::MATCH,
+            Etna::Clients::Magma::AttributeType::MATRIX,
+            Etna::Clients::Magma::AttributeType::TABLE,
+          ]
+
+          @valid_validation_types = [
```

Comment:
So by the way, you can actually generate this programmatically, any place that we're using enums like this, by creating a module and adding it to the enums like this:

```
module Enum
  def self.included(cls)
    cls.instance_eval do
      def entries
        self.constants.map { |c| self.const_get(c) }.select { |v| v.instance_of?(self) }
      end
    end
  end
end
```

Then just  `include Enum` and use `Etna::Clients::Magma::AttributeType.entries.select { |c| criteriahere }`